### PR TITLE
bug fix 10408: int for N in Poisson Conf throws error

### LINKED ISSUE #10408: poisson_confidence_interval with kraft-bourrows-nousek fails with integer input
### Description

The routines for Poisson confidence intervals are made for the Poisson distribution which is relevant when counting something (e.g. photons). Thus, it should work when an integer number
is passed in. 

### Expected behavior

Calculate statistics

### Actual behavior

For N > 100, i.e. when the calculation is run with mpmath, (for N< 100 scipy is used if available), an error is raised when n is an integer. It does work though if n is a float.

### Steps to Reproduce

```
In [1]: from astropy.stats import poisson_conf_interval                                               

In [2]: poisson_conf_interval(n=160, background=154.543, confidence_level=.95, interval='kraft-burrows
   ...: -nousek')                                                                                     
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-bfdb3197612a> in <module>
----> 1 poisson_conf_interval(n=160, background=154.543, confidence_level=.95, interval='kraft-burrows-nousek')

~/code/astropy/astropy/utils/decorators.py in wrapper(*args, **kwargs)
    533                     warnings.warn(message, warning_type, stacklevel=2)
    534 
--> 535             return function(*args, **kwargs)
    536 
    537         return wrapper

~/code/astropy/astropy/stats/funcs.py in poisson_conf_interval(n, interval, sigma, background, confidence_level)
    751         if np.any(background < 0):
    752             raise ValueError('Background must be >= 0.')
--> 753         conf_interval = np.vectorize(_kraft_burrows_nousek,
    754                                      cache=True)(n, background, confidence_level)
    755         conf_interval = np.vstack(conf_interval)

~/anaconda3/envs/astropy/lib/python3.8/site-packages/numpy/lib/function_base.py in __call__(self, *args, **kwargs)
   2089             vargs.extend([kwargs[_n] for _n in names])
   2090 
-> 2091         return self._vectorize_call(func=func, args=vargs)
   2092 
   2093     def _get_ufunc_and_otypes(self, func, args):

~/anaconda3/envs/astropy/lib/python3.8/site-packages/numpy/lib/function_base.py in _vectorize_call(self, func, args)
   2159             res = func()
   2160         else:
-> 2161             ufunc, otypes = self._get_ufunc_and_otypes(func=func, args=args)
   2162 
   2163             # Convert args to object arrays first

~/anaconda3/envs/astropy/lib/python3.8/site-packages/numpy/lib/function_base.py in _get_ufunc_and_otypes(self, func, args)
   2119 
   2120             inputs = [arg.flat[0] for arg in args]
-> 2121             outputs = func(*inputs)
   2122 
   2123             # Performance note: profiling indicates that -- for simple

~/code/astropy/astropy/stats/funcs.py in _kraft_burrows_nousek(N, B, CL)
   1273                                  'large.')
   1274     if HAS_MPMATH:
-> 1275         return _mpmath_kraft_burrows_nousek(N, B, CL)
   1276 
   1277     raise ImportError('Either scipy or mpmath are required.')

~/code/astropy/astropy/stats/funcs.py in _mpmath_kraft_burrows_nousek(N, B, CL)
   1179     from mpmath import mpf, factorial, findroot, fsum, power, exp, quad
   1180 
-> 1181     N = mpf(N)
   1182     B = mpf(B)
   1183     CL = mpf(CL)

~/anaconda3/envs/astropy/lib/python3.8/site-packages/mpmath/ctx_mp_python.py in __new__(cls, val, **kwargs)
     75         else:
     76             v = new(cls)
---> 77             v._mpf_ = mpf_pos(cls.mpf_convert_arg(val, prec, rounding), prec, rounding)
     78             return v
     79 

~/anaconda3/envs/astropy/lib/python3.8/site-packages/mpmath/ctx_mp_python.py in mpf_convert_arg(cls, x, prec, rounding)
     94                 return a
     95             raise ValueError("can only create mpf from zero-width interval")
---> 96         raise TypeError("cannot create mpf from " + repr(x))
     97 
     98     @classmethod

TypeError: cannot create mpf from 160
```
Note that passing a float works:
```
In [7]: poisson_conf_interval(n=160., background=154.543, confidence_level=.95, interval='kraft-burrow
   ...: s-nousek')                                                                                    
Out[7]: 
array([[ 0.        ],
       [30.30454909]])
```

and so does an integer with a smaller number:
```
In [8]: poisson_conf_interval(n=16, background=15.4543, confidence_level=.95, interval='kraft-burrows-
   ...: nousek')                                                                                      
Out[8]: 
array([[ 0.        ],
       [10.02142828]])
```

### System Details

macOS-10.13.6-x86_64-i386-64bit
Python 3.8.2 (default, Mar 26 2020, 10:43:30) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
Numpy 1.18.1
astropy 4.2.dev24+g441486c8c
Scipy 1.4.1

### PR DESCRIPTION
**Pull request by Iamsoto**
### Description

A Bug fix for issue #10408 

**The Problem**
Ints for parameter N appear to break `funcs.poisson_conf_interval(... interval='kraft-burrows-nousek')`

**Why?** 
`N` is type numpy.int64 when passed to `def _mpmath_kraft_burrows_nousek(N, B, CL)`
and mpmath.mpf() cannot convert from numpy.int64 so we get the error in the aforementioned issue. Though it appears mpmath.mpf can convert from numpy.float32. 

This happens because `numpy.vectorize()`  in `funcs.poisson_conf_interval(...)` passes dtype objects to its function (numpy.int64, numpy.float 32 etc... )

**My Solution**
In `def _mpmath_kraft_burrows_nousek(N, B, CL)` I check to see if the parameters are from the numpy module, if they are I convert them to their pythonic equivalencies. 

I also updated the doc string on the appropriate functions to make it clear the functions take python datatypes along with the appropriate numpy equivalency. 

**TL;DR**
Ints for parameter N now work in `funcs.poisson_conf_interval(... interval='kraft-burrows-nousek')`

Fixes #10408

### CODE REVIEW

File: astropy/stats/funcs.py [hunk 1]
```diff
@@ -1178,6 +1178,14 @@ def _mpmath_kraft_burrows_nousek(N, B, CL):
...
     from mpmath import mpf, factorial, findroot, fsum, power, exp, quad
 
+    if type(N).__module__ == np.__name__:
+        N = N.item()
+
+    if type(B).__module__ == np.__name__:
+        B = B.item()
+
+    if type(CL).__module__ == np.__name__:
+        CL = CL.item()
```
Comment (hamogu) [inline astropy/stats/funcs.py:1188]: Would it not be easier to just do `N = mpf(float(N))`? That way it would work for anything that can be converted to a float. Checking the `__module__ of the type is a great idea I would have never thought of, but, unless really needed, is also pretty complicated.

Comment (Iamsoto) [reply astropy/stats/funcs.py:1188]: I guess all mpf does is convert the input into a float. So yeah that should work. I'm not sure why I didn't think of this XD

**Comment by hamogu** [comment]
Ok, I think this is good except for a few very small things:

- The whitespace error that pep8bot points out.
- Can you put in a changelog entry under bug fixes, statistic, milestone 4.1.1?
- @larrybradley : Please let us know if you have any objections. If I don't hear from you, I'll approve and merge this since it's a fairly small change.

**Comment by Iamsoto** [comment]
Sure thing,

Could you please briefly explain to me where the changelog is and how to add to it?
I can't seem to find the information on the site.

**Comment by hamogu** [comment]
See here: https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#add-a-changelog-entry

You need to edit this file: https://github.com/astropy/astropy/blob/master/CHANGES.rst It's sorted by milestones, but there is no section 4.1.1 yet. If you look into that file, you'll see the format of the entries and it's pretty obvious how to write one.

@pllim @bsipocz : This PR is milestoned to 4.1.1 but 4.1 is still unreleased (according to the CHANGES.rst on master) and thus there is no section 4.1.1 in the file. How we we add a changelog entry? Do we need to wait a could of days until 4.1 is officially out? Or does it go into 4.1 (unreleased)?

**Comment by pllim** [comment]
FWIW, I am trying to add the requested section in #10447 but you can't pick it up until it has been merged. FYI.

**Comment by pllim** [comment]
Not a blocker for the PR but just in case you are unaware, the email you use to create the Git commits are not registered to your GitHub account, @Iamsoto .

**Comment by Iamsoto** [comment]
Thanks for the help guys, 

Looks like my upstream is set to the master branch... It appears I don't have a 4.1.1 section. I tried to make an edit in 4.1 (unreleased). Guess it doesn't like that

**Comment by pllim** [comment]
@Iamsoto , you might not have picked up the latest changes or somehow you are pulling from your fork's master, and not astropy's master. Also, that merge commit needs to go.

Before you proceed below, please make a backup branch if you have any doubts at all, just in case.

Please try the following (replace `upstream` with whatever you name the remote that points here, not your fork, if applicable):

```
git fetch upstream master
git rebase upstream/master
git status
git log
```

Make sure status is clean and your log history makes sense. Then you will need to force push.

```
git push -f origin bugfix_10408_poisson_conf
```

Hope this helps!

**Comment by pllim** [comment]
If things go well, you will see the new 4.1.1 section and you can move your entry over there.

**Comment by Iamsoto** [comment]
Hey, I'm sorry about the mixup and I really appreciate your guys' help! 
I removed the merge with a  reset --hard HEAD~2 
then followed the steps you told me. Then edited the changelog. 

Hopefully everything is good now.

**Comment by pllim** [comment]
Bot is happy. 🤖 😄 

Thanks for your patience!

**Comment by hamogu** [comment]
Will merge, when CI passes.

**Comment by astrofrog** [comment]
@hamogu @pllim - just to check, is there a reason this can't be included in 4.0.2? Or is the bug fix not needed there?

**Comment by hamogu** [comment]
It could be included there and the bugfix in needed all the way back to where this function was implemented (around 1.0 I think). Do we have rules on how to decide on a milestone somewhere? I have never set a milestone and this is the first time I ever merged a PR for astropy.

**Comment by astrofrog** [comment]
@hamogu - I'm not sure if the rule is written down anywhere but basically if it's a bug fix, it should be milestoned as the next LTS bugfix (in this case 4.0.2) unless it can only be applied to the latest version in which case it would be 4.1.1 as you did here. If 4.2 was out, it means the milestone would either be 4.0.2 or 4.2.1. In any case, it's easily fixable here - the milestone needs to be changed to 4.0.2, and we just need a small follow up PR also milestoned 4.0.2 that moves the changelog entry. Would you be able to do this?

**Comment by pllim** [comment]
I changed the milestone. FYI.

**Comment by bsipocz** [comment]
FYI: no need to follow-up PR for changelog cleanups, the release procedure can and should be able to deal with syncing up the changelog to the actual reality of the branch.
